rfd/0193-access-control-decision-api.md

```diff
@@ -0,0 +1,345 @@
+
+  // ...
+}
+
+message EvaluateSSHAccessRequest {
+  RequestMetadata metadata = 1;
+
+  SSHIdentity ssh_identity = 2;
+
+  Resource server = 3;
+
+  string login = 4;
+}
+
+message EvaluateSSHAccessResponse {
+  oneof decision {
+    SSHAccessPermit permit = 1;
+    SSHAccessDenial denial = 2;
+  }
+}
+
+message SSHAccessPermit {
+  PermitMetadata metadata = 1;
+
+  repeated string logins = 2;
+
+  bool forward_agent = 3;
+
+  google.protobuf.Duration max_session_ttl = 4;
+
+  bool port_forwarding = 5;
+
+  int64 client_idle_timeout = 6;
+
+  bool disconnect_expired_cert = 7;
+
+  repeated string bpf = 8;
+
+  bool x11_forwarding = 9;
+
+  int64 max_connections = 10;
+
+  // ... (there's a lot more that needs to go here)
+}
+
+message SSHAccessDenial {
+  DenialMetadata metadata = 1;
+
+  // ... (this may eventually support MFA flows, etc.)
+}
+
+message SSHIdentity {
+  // ... (will contain repr of standard teleport ssh cert identity)
+}
+```
+
+With the above API, most server access enforcement logic can be reworked to respect the PDP/PEP boundary with minimal changes.
+`AccessChecker.EnhancedRecordingSet()` becomes `permit.Bpf`, `AccessChecker.GetHostSudoerts(server)` becomes
+`permit.HostSudoers`, etc.
+
+
+### Cross-Version Compatibility
+
+One of the core problems that we need to solve is how to ensure that version/feature mismatches between PDP and PEP are easy to
+handle in an intuitive and correct manner. Failure to correctly handle cross-version compat can lead to serious security issues,
+and the current system of injecting wildcard deny rules at the caching layer is coarse, leads to false-positive denials, and is
+cumbersome to implement correctly.
+
+For most teleport API interactions where cross-version compat matters, we simply share the version
+of the client with the server and vise-versa and make decisions based on that. This quickly becomes unwieldy, however,
+when features start getting backported to preceeding major versions. Some APIs now use feature flags, which are much easier to work
+with and backport, and are generally quite reliable. However, using either poses problems for a PDP API because some decisions need
+to be made in contexts where the version of the PEP may not be known with 100% certainty. For example, decisions made on the proxy will
+rely on details from the agent heartbeat, which may be outdated.
+
+To solve this, permits will contain "feature assertions", which will specify which features *must* be available for the permit to be
+properly enforced. Enforcing agents will then reject otherwise allowable actions if the permit contains one or more unknown features.
+Feature assertions will be represented as a repeated enum for efficient encoding. Because feature assertions can be appended to permits
+on a case-by-case basis, we minimize false-positives (instances where access gets denied due to an outdated enforcement point even though
+the specific access did not merit the use of any new controls).
```

Comment:
Can we include a concrete example in this section? I'll share one to check my understanding.

Let's say we add new feature where a role/policy can allow SSH port forwarding of only server ports or only client ports, and older PEPs only understand a binary allow/deny of port forwarding at all.
- we'd add a feature assertion `FEATURE_ASSERTION_ENHANCED_PORT_FORWARDING`
- the PDP would set this assertion only if the decision was to allow exactly one of server or client port forwarding
- the older PEP (ssh agent) would have to completely disregard the permit if it doesn't understand the new permit, meaning it could not allow SSH access at all

Reply:
>A possible solution is to move the feature assertions into each individual option.

I don't think we need to do that to get granularity, the features are already meant to be granular. There's little difference between a "global" _ENHANCED_PORT_FORWARDING vs a "localized" one, but the latter means the features are spread out through the definitions and more messages need to change.

There are also compatibility problems with adding features to a message that didn't carry them before.

Reply:
The difference between global and localized feature protection lies in how an agent handles unsupported features. If the agent is running an older version that doesn't support a newly introduced feature, a global enum representing the feature will be unknown to it.

Consider this scenario:  
- The control plane is running version `vx.y.z`.  
- The agent is running version `vx.y.(z-2)`.  
- In version `vx.y.(z-1)`, a new feature—enhanced port forwarding—was introduced, along with its corresponding feature protection in the protocol.

Since the agent is running a version older than the one where the feature flag was added, it will treat the new enum value as just an unrecognized number with no context about the associated feature. As @nklaassen pointed out, in such cases, the agent’s only safe course of action is to block the entire protocol—for example, preventing SSH access to the server. This precaution is necessary because the unknown flag could control any critical SSH feature, such as port forwarding, forward agent, allowed commands... Without understanding the feature, the secure approach is to deny access entirely.

If features are localized, however, the unknown feature protection would be specific to port forwarding. In this case, the agent could make a more targeted decision: it would allow SSH connections but deny **all** port forwarding requests, since it cannot interpret the flag value.

In certain scenarios, ensuring correctness requires blocking SSH access, which makes a global flag necessary. With the decision now centralized, the Policy Decision Point (PDP) must either have full knowledge of the features supported by each agent version or we need to design a system that enables the agent to make appropriate decisions based on feature protection flags.

Previously, the authentication server handled these decisions by blocking protocol access when needed—e.g., downgrading roles and applying deny labels like `{'*': '*'}`. However, this approach was not scalable, and there have been instances in the past where features were introduced without properly restricting access when agents did not support them.

Reply:
I think the intent here is to only add the feature enum to the permit if the feature would be relevant. The agent would then deny the request if any feature it doesn't know about is present, like you said.

This is the same as a localized feature, except the control plane code makes the decision of whether the feature applies or not. The end result should be the same, but with simpler agent code.

Reply:
Discussed out-of-band with Tiago at length - I'm not fully convinced a localized feature is necessary, but if so I think there are alternative designs that can make the fallback logic come from the control plane as well (instead of having to code the fallback at the agent itself).

My 2c is that new features shouldn't be on by default and should only be set by the control plane if the feature is relevant. This means that users need to enable the features explicitly, which also means they would know they've done so and can disable the feature while a rollout is in progress. (I'll admit reality isn't always this smooth though.)

Happy to share a fallback-from-control-plane design idea if it comes to that.

Reply:
There's another option I don't think we've discussed here, which is that on a case-by-case basis the global feature flag may not be necessary. In this port forwarding example, the PDP could set the legacy PortForwarding boolean to false (if any port forwarding option is denied), and set the new enhanced port forwarding options correctly, and that way old PEPs will fail-safe and never allow any port forwarding. I suspect we could do something similar for most new access control features

Reply:
Yes, I think this falls here:

> - If possible, implement new features s.t. it is safe for older PEPs to ignore them

Features are a last resort - if whatever we want can be done as you suggest than it's a much better way.

Reply:
The PDP could send an informational message to the agent when fallbacks are in effect. Would that solve this concern?

Reply:
Doesn't that mean PDP would need to be aware of all possible version features, or that we would have to send that information each time? If it's the latter, wouldn't that be essentially the same as directly sending the feature as enum/json payload?

Reply:
Updated the RFD with more concrete examples that I hopefully flush it out a bit more.

While I see why per-control feature assertions do have nice qualities, I'm wary to recommend them as I think they would introduce a *lot* of extra noise.  The permits are already going to end up being fairly large/complex, and adding dedicated assertion fields for every individual control just in case we decide to make it more granular in the future strikes me as iffy.  I think that the idea of leaving the old field in the "off" position and introducing a new one when a single control gets split into more granular sub-controls is likely going to be much cleaner in the long run, even if slightly less expressive.  After all, we don't do that kind of thing very often, and we can always deprecate the old field after a few major versions to keep the permit from growing indefinitely.